Valid code is not considered valid by the type checkers

## What was wrong?

The signature of `is_same_address` requires both arguments to be `eth_typing.AnyAddress`, but actually, the source code of the function allows any string to be provided.

It would be nice if the typing signature of the function became less restrictive because it is a bit cumbersome to wrap plain strings into proper types just to satisfy the type checker while adding `type: ignore` is not the best solution either.

### Code that produced the error

```py
from eth_utils import is_same_address
from eth_typing import HexAddress, HexStr

print(is_same_address("0x0000000000000000000000000000000000000000", "0x0000000000000000000000000000000000000000"))
print(is_same_address(HexAddress(HexStr("0x0000000000000000000000000000000000000000")), HexAddress(HexStr("0x0000000000000000000000000000000000000000"))))
```

### Full error output 

```sh
% python f.py
True
True
% mypy f.py
f.py:4: error: Value of type variable "AnyAddress" of "is_same_address" cannot be "str"
Found 1 error in 1 file (checked 1 source file)
```

### Expected Result

```sh
% python f.py
True
True
% mypy f.py
Success: no issues found in 1 source file
``` 

### Environment

```sh
% python -m eth_utils
Python version:
3.9.9 (main, Jan 10 2022, 09:25:51)
[Clang 13.0.0 (clang-1300.0.29.30)]

Operating System: macOS-12.2.1-x86_64-i386-64bit

pip freeze result:
cytoolz==0.11.2
eth-hash==0.3.2
eth-typing==3.0.0
eth-utils==2.0.0
mypy==0.942
mypy-extensions==0.4.3
tomli==2.0.1
toolz==0.11.2
typing_extensions==4.1.1
``` 

And the `mypy.ini` for the reference:

```
% cat mypy.ini
[mypy]
ignore_missing_imports = True

check_untyped_defs = True
disallow_any_generics = True
disallow_untyped_defs = True
follow_imports = silent
strict_optional = True
warn_redundant_casts = True
warn_unused_ignores = True

plugins = pydantic.mypy

[mypy-app.db.migrations.*]
ignore_errors = True
```

## How can it be fixed?

Update the type signature of `is_same_address` and maybe other similar functions to accept `Union[eth_utils.AnyAddress, typing.AnyStr]` instead.

--- a/eth_utils/address.py
+++ b/eth_utils/address.py
@@ -110,7 +110,9 @@ def is_canonical_address(address: Any) -> bool:
     return cast(bool, is_equal)
 
 
-def is_same_address(left: AnyAddress, right: AnyAddress) -> bool:
+def is_same_address(
+    left: Union[AnyAddress, str, bytes], right: Union[AnyAddress, str, bytes]
+) -> bool:
     """
     Checks if both addresses are same or not.
     """

--- a/newsfragments/289.bugfix.rst
+++ b/newsfragments/289.bugfix.rst
@@ -1,0 +1,1 @@
+Update types in `is_same_address` to accept `AnyAddress`, `str`, or `bytes`.

